### PR TITLE
Introduce Parameter

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

File: include/deepworks/shape.hpp [hunk 1]
```diff
@@ -1,9 +1,19 @@
 #pragma once
 
 #include <vector>
+#include <ostream>
 
 namespace deepworks {
 
 using Shape = std::vector<int>;
 
+inline std::ostream& operator<<(std::ostream& os, Shape shape) {
```
Comment (TolyaTalamanov) [inline include/deepworks/shape.hpp:10]: `const Shape&`

**Comment by TolyaTalamanov** [comment]
@l-bat @RodionovF @KSTconsult @apertovs Guys, have a look

File: src/runtime/cpu/layers/cpulinear.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    const auto& in_shape  = input.shape();
+    const auto& out_shape = output.shape();
+
+    deepworks::CPULinearForward({input.data() , in_shape[0] , in_shape[1]},
+                                {weight.data(), w_shape[0]  , w_shape[1]},
+                                {output.data(), out_shape[0], out_shape[1]});
+
+    // FIXME: Signature is incorrect.
+    // Can AddBias works with Vector (single dimmension) ???
+    //deepworks::CPULinearAddBias({bias.data() ,  1 , b_shape[0]},
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpulinear.cpp:39]: Uncomment damn, it works because bias is zero value by default

File: src/runtime/cpu/layers/cpulinear.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    const auto& w_shape   = weight.shape();
+    const auto& b_shape   = bias.shape();
+    const auto& in_shape  = input.shape();
+    const auto& out_shape = output.shape();
+
+    deepworks::CPULinearForward({input.data() , in_shape[0] , in_shape[1]},
+                                {weight.data(), w_shape[0]  , w_shape[1]},
+                                {output.data(), out_shape[0], out_shape[1]});
+
+    // FIXME: Signature is incorrect.
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpulinear.cpp:37]: @KSTconsult Let's discuss it locally

File: tests/unit/test_utils.hpp [hunk 4]
```diff
@@ -20,4 +20,5 @@ inline void AssertTensorEqual(const deepworks::Tensor& actual, const deepworks::
         ASSERT_FLOAT_EQ(expected_p[i], actual_p[i]);
     }
 }
+
```
Comment (TolyaTalamanov) [inline tests/unit/test_utils.hpp:23]: Unnecessary changes

**Comment by TolyaTalamanov** [comment]
@RodionovF @l-bat @KSTconsult @apertovs Have a look, please

File: include/deepworks/model.hpp [hunk 5]
```diff
@@ -20,6 +21,7 @@ class Model {
           Layers      & layers();
 
     Layer getLayer(const std::string& name);
+    Parameters& params();
```
Comment (TolyaTalamanov) [inline include/deepworks/model.hpp:24]: @RodionovF FYI: 

```cpp

SGDOptimizer opt(model.params(), 0.001);
opt.step()
```
as possible API

File: include/deepworks/parameter.hpp [hunk 6]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <deepworks/tensor.hpp>
+
+namespace deepworks {
+
+class Parameter {
```
Comment (TolyaTalamanov) [inline include/deepworks/parameter.hpp:7]: @RodionovF FYI

File: include/deepworks/shape.hpp [hunk 7]
```diff
@@ -1,9 +1,19 @@
...
+    os << "[";
+    for (auto sh : shape) {
+        os << sh << " ";
```
Comment (l-bat) [inline include/deepworks/shape.hpp:13]: Let's add delimiter: 
```suggestion
        os << sh << ," ";
```

File: src/runtime/cpu/layers/cpulinear.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
+#include "runtime/cpu/layers/cpulinear.hpp"
+#include "runtime/cpu/kernels/kernels.hpp"
+
+#include <iostream>
```
Comment (l-bat) [inline src/runtime/cpu/layers/cpulinear.cpp:4]: remove

File: include/deepworks/parameter.hpp [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    Parameter(Tensor&& data, bool is_trainable=true);
+
+    const Tensor& data() const;
+          Tensor& data();
+
+    const Tensor& grad() const;
+          Tensor& grad();
+
+    void train(bool mode);
+    bool is_trainable() const;
```
Comment (RodionovF) [inline include/deepworks/parameter.hpp:18]: Do we need to make the flag a tensor too?

now we turn off/on the whole layer?

Comment (TolyaTalamanov) [reply include/deepworks/parameter.hpp:18]: No, because for instance: `Linear` layer has two parameterers for `weight` and `bias`, so we can freeze only `weight`

File: include/deepworks/shape.hpp (hunk 7 above)
Comment (TolyaTalamanov) [reply include/deepworks/shape.hpp:13]: So, then it will be `[, 3, 5, 5]`

Comment (TolyaTalamanov) [reply include/deepworks/shape.hpp:13]: BTW comma outside of string :(

File: src/expression/nn.cpp [hunk 10]
```diff
@@ -1,27 +1,44 @@
...
+#include "util/assert.hpp"
 
 deepworks::Linear::Linear(int units, std::string name)
     : BaseOp<deepworks::Linear>(deepworks::LayerInfo(std::move(name), "Linear")) {
     m_info.impl().attrs["units"] = units;
 }
 
-deepworks::Shape deepworks::Linear::output_shape(const deepworks::Shape& input) {
-    return {input[0], m_info.impl().attrs["units"].get<int>()};
+deepworks::Shape deepworks::Linear::output_shape(const deepworks::Shape& in_shape) {
```
Comment (TolyaTalamanov) [inline src/expression/nn.cpp:14]: Possible it should be done at separate function, I'll put FIXME

Comment (TolyaTalamanov) [reply src/expression/nn.cpp:14]: Done

File: src/runtime/cpu/layers/cpulinear.cpp (hunk 8 above)
Comment (TolyaTalamanov) [reply src/runtime/cpu/layers/cpulinear.cpp]: Done

File: include/deepworks/shape.hpp (hunk 7 above)
Comment (TolyaTalamanov) [reply include/deepworks/shape.hpp]: Fixed

File: tests/unit/test_utils.hpp (hunk 4 above)
Comment (TolyaTalamanov) [reply tests/unit/test_utils.hpp]: Done

File: src/runtime/cpu/kernels/kernels.cpp [hunk 11]
```diff
@@ -1,15 +1,21 @@
 #include "runtime/cpu/kernels/kernels.hpp"
 
+// FIXME: It can something general like:
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/kernels/kernels.cpp:3]: @KSTconsult Just put a reminder, it future it can be decomposed to `Mul` and `Add` operations

File: include/deepworks/shape.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply include/deepworks/shape.hpp]: Done

**Comment by TolyaTalamanov** [comment]
@RodionovF @apertovs @l-bat @KSTconsult Have a look once again and let's merged it!

**Review by l-bat — APPROVED** [review]

File: src/expression/nn.cpp [hunk 12]
```diff
@@ -1,27 +1,47 @@
...
+#include "util/assert.hpp"
 
 deepworks::Linear::Linear(int units, std::string name)
     : BaseOp<deepworks::Linear>(deepworks::LayerInfo(std::move(name), "Linear")) {
     m_info.impl().attrs["units"] = units;
 }
 
-deepworks::Shape deepworks::Linear::output_shape(const deepworks::Shape& input) {
-    return {input[0], m_info.impl().attrs["units"].get<int>()};
+void deepworks::Linear::init(const Shape& in_shape) {
```
Comment (TolyaTalamanov) [inline src/expression/nn.cpp:13]: @apertovs FYI, now it's computing in separate method

File: src/runtime/cpu/kernels/kernels.cpp (hunk 11 above)
Comment (TolyaTalamanov) [reply src/runtime/cpu/kernels/kernels.cpp:3]: Decomposed in this PR: https://github.com/NeuralDeepWorks/deepworks/pull/23

**Review by RodionovF — APPROVED** [review]

File: src/model/parameter.cpp [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+const deepworks::Tensor& deepworks::Parameter::grad() const {
+    return m_impl->m_grad;
+}
+
+deepworks::Tensor& deepworks::Parameter::grad() {
+    return m_impl->m_grad;
+}
+
+void deepworks::Parameter::train(bool mode) {
```
Comment (apertovs) [inline src/model/parameter.cpp:33]: Should we rename this method to set_mode?
Or we can have eval() and train() methods to switch the mode.

**Review by apertovs — APPROVED** [review]
Looks good to me.

File: src/model/parameter.cpp (hunk 13 above)
Comment (TolyaTalamanov) [reply src/model/parameter.cpp:33]: I believe `set_mode` can be used directly for entire `dw::Model`, when this one used only for freeze specific `param`
